### PR TITLE
adding rc_flags for reagent containers behavior

### PR DESCRIPTION
**Pull request by Buyrcsp2**
A-Chemistry, C-Feature, C-Rework
## About the PR 
Added rc_flags, the original rc_flags renamed to rc_desc_flags. Currently adding CAN_TRANSFER, CAN_RECEIVE, CAN_SPLASH (self-explanatory)
could cause unwanted behaviors on objects without rc_flags, also made space cola use this for an example of the usage of the flags.

## Why's this needed? 
Because it can make people make certain behaviors for certain object or reagent containers without too much extra code

### CODE REVIEW

File: code/WorkInProgress/MarqStuff.dm [hunk 1]
```diff
@@ -444,7 +444,7 @@
 			return 1
 		else
 			var/obj/item/I = target
-			if (istype(I) && I.is_open_container() == 1 && I.reagents)
+			if (istype(I) && I.is_can_receive() == 1 && I.reagents) //hello if you're seeing this then my code is approved, you'll need to use is_can_receive instead of is_open_container, it's the same thing with different names
```
Comment (ZeWaka) [inline code/WorkInProgress/MarqStuff.dm:447]:
```suggestion
			if (istype(I) && I.is_can_receive() == 1 && I.reagents)
```
no

File: _std/defines/chemistry.dm [hunk 2]
```diff
@@ -38,4 +38,7 @@
 #define THRESHOLD_UNDER 0
 #define THRESHOLD_OVER 1
 #define THRESHOLD_INIT THRESHOLD_UNDER
-
+//What can be done with a reagent_container
+#define CAN_TRANSFER (1<<0) //can transfer to containers
+#define CAN_RECEIVE  (1<<1) //can receive reagents from containers
+#define CAN_SPLASH  (1<<2) //can splash reagent from this
```
Comment (ZeWaka) [inline _std/defines/chemistry.dm:44]: doc comments

File: code/atom.dm [hunk 3]
```diff
@@ -8,6 +8,8 @@
 	plane = PLANE_DEFAULT
 	var/level = 2
 	var/flags = FPRINT
+	var/rc_desc_flag = RC_SPECTRO | RC_FULLNESS | RC_VISIBLE
+	var/rc_flags = null
```
Comment (ZeWaka) [inline code/atom.dm:12]: doc comments

File: code/atom.dm [hunk 4]
```diff
@@ -206,8 +208,14 @@
...
-	proc/is_open_container()
-		return flags & OPENCONTAINER
+	proc/is_can_receive()
+		return rc_flags & CAN_RECEIVE
+
+	proc/is_can_transfer()
+		return rc_flags & CAN_TRANSFER
+
+	proc/is_can_splash()
+		return rc_flags & CAN_SPLASH
```
Comment (ZeWaka) [inline code/atom.dm:218]: i doubt that comment is still valid.

also, doc comments

Comment (ZeWaka) [inline code/atom.dm:218]: `is_can` sounds incredibly wordy and poor english

File: code/modules/chemistry/tools/food_and_drink.dm [hunk 5]
```diff
@@ -615,111 +615,6 @@ ABSTRACT_TYPE(/obj/item/reagent_containers/food/snacks)
 		consumer.urine += 0.1
 		eat_twitch(consumer)
 
-	//bleck, i dont like this at all. (Copied from chemistry-tools reagent_containers/glass/ definition w minor adjustments)
-	// still copy paste btw
-	afterattack(obj/target, mob/user , flag)
-		if (is_sealed)
-			boutput(user, "<span class='alert'>[src] is sealed.</span>")
```
Comment (ZeWaka) [inline code/modules/chemistry/tools/food_and_drink.dm:622]: why is this deleted

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/atom.dm (hunk 4 above)
Comment (Buyrcsp2) [reply code/atom.dm:218]: will a simple can_() work?

File: code/modules/chemistry/tools/food_and_drink.dm (hunk 5 above)
Comment (Buyrcsp2) [reply code/modules/chemistry/tools/food_and_drink.dm:622]: i moved the code to obj/item/reagent_containers and thus make it inherit the code, no need for 2 blocks of the same code

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 6]
```diff
@@ -331,7 +332,7 @@
 		src.add_fingerprint(user)
 		src.enter_bathtub(target)
 
-	is_open_container()
+	can_receive() //OPENCONTAINER renamed to CAN_RECEIVE.
```
Comment (ZeWaka) [inline code/WorkInProgress/DrMelonsStuff.dm:335]:
```suggestion
	can_receive()
```
i don't care about a flag that doesn't exist anymore

File: code/atom.dm [hunk 7]
```diff
@@ -203,11 +205,15 @@
 
 /**
   * Convenience proc to see if a container is open for chemistry handling
```
Comment (ZeWaka) [inline code/atom.dm:207]: all of this too.

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Comment by Tarmunora** [comment]
scope creep, but would probably be good to standardize mousedrop transfers into this flags/procs schema at the same time

File: _std/defines/chemistry.dm [hunk 8]
```diff
@@ -38,4 +38,7 @@
 #define THRESHOLD_UNDER 0
 #define THRESHOLD_OVER 1
 #define THRESHOLD_INIT THRESHOLD_UNDER
-
+//What can be done with a reagent_container
+#define CAN_TRANSFER (1<<0) //can transfer to other containers
+#define CAN_RECEIVE  (1<<1) //can receive reagents from other containers
+#define CAN_SPLASH  (1<<2) //can splash reagent inside this container
```
Comment (ZeWaka) [inline _std/defines/chemistry.dm:44]:
```suggestion
/// can transfer to other containers
#define CAN_TRANSFER (1<<0)
/// can receive reagents from other containers
#define CAN_RECEIVE (1<<1)
/// can splash reagent inside this container
#define CAN_SPLASH (1<< 2)
```

File: code/atom.dm [hunk 9]
```diff
@@ -202,12 +204,16 @@
 		return null
 
 /**
-  * Convenience proc to see if a container is open for chemistry handling
-	*
-  * * returns true if open, false if closed
+  * Convenience procs
```
Comment (ZeWaka) [inline code/atom.dm:207]: as in, remove the whole comment

File: code/atom.dm [hunk 10]
```diff
@@ -202,12 +204,16 @@
...
 	*/
-	proc/is_open_container()
-		return flags & OPENCONTAINER
+	proc/can_receive()
+		return rc_flags & CAN_RECEIVE //checks for CAN_RECEIVE
```
Comment (ZeWaka) [inline code/atom.dm:210]: useless comment, please document the actual proc and what it's used for/why

File: code/modules/chemistry/Chemistry-Tools.dm [hunk 11]
```diff
@@ -392,19 +345,19 @@ proc/ui_describe_reagents(atom/A)
...
+						boutput(user, "You pour [transferred] units of the [W]'s contents into the [src].")
 					else
-						boutput(user, "<span class='alert'>The [src.name] is full.</span>")
+						boutput(user, "<span class='alert'>The [src] is full.</span>")
 				else
-					boutput(user, "The [W.name] is empty.")
+					boutput(user, "The [W] is empty.")
 			else
-				boutput(user, "You need to slice open the [W.name] first!")
+				boutput(user, "You need to slice open the [W] first!")
```
Comment (ZeWaka) [inline code/modules/chemistry/Chemistry-Tools.dm:360]: What's with these changes? Now there's going to be `The the beaker`s

File: code/modules/chemistry/Chemistry-Tools.dm [hunk 12]
```diff
@@ -12,10 +12,11 @@ ABSTRACT_TYPE(/obj/item/reagent_containers)
 	icon_state = null
 	w_class = W_CLASS_TINY
 	flags = FPRINT | TABLEPASS | SUPPRESSATTACK
-	var/rc_flags = RC_VISIBLE | RC_FULLNESS | RC_SPECTRO
 	tooltip_flags = REBUILD_SPECTRO | REBUILD_DIST
+	rc_desc_flag = RC_SPECTRO | RC_VISIBLE | RC_FULLNESS
 	var/amount_per_transfer_from_this = 5
 	var/initial_volume = 50
+	var/splash_all_contents = 1
```
Comment (ZeWaka) [inline code/modules/chemistry/Chemistry-Tools.dm:19]: should this be a 1 or a TRUE

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/WorkInProgress/DrMelonsStuff.dm [hunk 13]
```diff
@@ -145,12 +146,13 @@
 		else if (istype(over_object, /turf))
 			drain_bathtub(usr)
 			return
-		if (!(over_object.flags & ACCEPTS_MOUSEDROP_REAGENTS))
+		if (!(over_object.flags & ACCEPTS_MOUSEDROP_REAGENTS & can_transfer()))
```
[the remark this answers is not in the document]
Comment (Buyrcsp2) [reply code/WorkInProgress/DrMelonsStuff.dm:149]: forgot to remove that

**Review by ZeWaka — APPROVED** [review]

**Comment by ZeWaka** [comment]
@Buyrcsp2 You did absolutely zero testing for this PR - frankly unacceptable.  Reverted.
